### PR TITLE
skaffold init --force supports cases with 1 image and multiple builders

### LINKED ISSUE #3661: FR: Can `skaffold init --force` automatically select a builder if there are multiple builders and a single image detected?
skaffold version: 1.3.1

**Update:**
For example, I would like the following scenario to succeed in https://github.com/GoogleCloudPlatform/cloud-code-samples/tree/master/golang/go-hello-world:

```
$ SKAFFOLD_XXENABLEBUILDPACKSINIT=1 SKAFFOLD_XXENABLEBUILDPACKINIT=1 skaffold init --analyze | jq
{
  "builders": [
    {
      "name": "Docker",
      "payload": {
        "path": "Dockerfile"
      }
    },
    {
      "name": "Buildpacks",
      "payload": {
        "path": "go.mod",
        "builder": "heroku/buildpacks"
      }
    }
  ],
  "images": [
    {
      "name": "go-hello-world",
      "foundMatch": false
    }
  ]
}
$ SKAFFOLD_XXENABLEBUILDPACKSINIT=1 SKAFFOLD_XXENABLEBUILDPACKINIT=1 skaffold init --force
FATA[0000] unable to automatically resolve builder/image pairs; run `skaffold init` without `--force` to manually resolve ambiguities
```

In the event that there are multiple candidate builders and a single image detected, is it feasible for Skaffold to choose them for me according to some heuristic when `--force` is used? 

We could pick the highest precedence builder according to this list from https://github.com/GoogleContainerTools/skaffold/issues/3661#issuecomment-592732189:
1. Jib
2. Bazel
3. Dockerfile
4. Buildpack

**Original issue, disregard:**
For example, I would like the following scenario to succeed in https://github.com/GoogleCloudPlatform/cloud-code-samples/tree/master/golang/go-guestbook:

```
$ SKAFFOLD_XXENABLEBUILDPACKSINIT=1 SKAFFOLD_XXENABLEBUILDPACKINIT=1 SKAFFOLD_XXENABLEJIBINIT=1 skaffold init --analyze | jq
{
  "builders": [
    {
      "name": "Docker",
      "payload": {
        "path": "src/backend/Dockerfile"
      }
    },
    {
      "name": "Buildpacks",
      "payload": {
        "path": "src/backend/go.mod",
        "builder": "heroku/buildpacks"
      }
    },
    {
      "name": "Docker",
      "payload": {
        "path": "src/frontend/Dockerfile"
      }
    },
    {
      "name": "Buildpacks",
      "payload": {
        "path": "src/frontend/go.mod",
        "builder": "heroku/buildpacks"
      }
    }
  ],
  "images": [
    {
      "name": "go-guestbook-backend",
      "foundMatch": false
    },
    {
      "name": "go-guestbook-frontend",
      "foundMatch": false
    },
    {
      "name": "mongo",
      "foundMatch": false
    }
  ]
}
$ SKAFFOLD_XXENABLEBUILDPACKSINIT=1 SKAFFOLD_XXENABLEBUILDPACKINIT=1 SKAFFOLD_XXENABLEJIBINIT=1 skaffold init --force
FATA[0000] unable to automatically resolve builder/image pairs; run `skaffold init` without `--force` to manually resolve ambiguities
```

In the event that there are multiple candidate builders per image, is it feasible for Skaffold to choose them for me according to some heuristic when `--force` is used? In this example, `Docker` and `Buildpacks` builders were present. Skaffold could potentially choose `Docker` builders for me without prompting.

### PR DESCRIPTION
**Pull request by MarlonGamez**
Fixes #3661

**Description**
Allows for `skaffold init --force` to work in situations where it detects one image and multiple builders. Previously, the user would see the following
```
FATA[0000] unable to automatically resolve builder/image pairs; run `skaffold init` without `--force` to manually resolve ambiguities
```
Now, a builder will be chosen with the following priority: Docker > Jib > Bazel > Buildpacks
This ordering was decided upon from input on both #3661 and https://kubernetes.slack.com/archives/CABQMSZA6/p1602786450111700.

**Follow-up Work**
It would be nice to do something in the case of multiple images and builders. I've opened #4912 to talk about this.

### CODE REVIEW

**Comment by tejal29** [comment]
Can we update the init docs with the order of preference?

File: pkg/skaffold/initializer/build/resolve.go [hunk 1]
```diff
@@ -53,12 +53,47 @@ func (d *defaultBuildInitializer) resolveBuilderImages() error {
...
+		image := d.unresolvedImages[0]
+		choice := d.builders[0]
+		for _, builder := range d.builders {
+			if builderRank(builder) < builderRank(choice) {
+				choice = builder
+			}
+		}
+
+		d.builderImagePairs = append(d.builderImagePairs, BuilderImagePair{Builder: choice, ImageName: image})
+		d.unresolvedImages = util.RemoveFromSlice(d.unresolvedImages, image)
```
Comment (gsquared94) [inline pkg/skaffold/initializer/build/resolve.go:74]: should `unresolvedImages` be a `map` instead?

**Comment by MarlonGamez** [comment]
> Can we update the init docs with the order of preference?

@tejal29 It actually looks like there are no docs on the `--force` flag, so I can write something up for that. Should this be a separate PR?

File: pkg/skaffold/initializer/build/resolve.go (hunk 1 above)
Comment (MarlonGamez) [reply pkg/skaffold/initializer/build/resolve.go:74]: perhaps! Could you explain your thoughts on why it should be?

**Comment by tejal29** [comment]
> > Can we update the init docs with the order of preference?
> 
> @tejal29 how would you suggest documenting this without being too verbose? If we say something like "In the case of multiple builders being matched to a single image, the priority is Docker > Jib > Bazel > Buildpacks" I'm worried it might make things really long

Ah not the output. I meant the Init docs https://skaffold.dev/docs/pipeline-stages/init/. 
We don't have `--force` documented here. Upto you if you want to create another PR for this or same one.

File: pkg/skaffold/initializer/build/resolve.go (hunk 1 above)
Comment (tejal29) [inline pkg/skaffold/initializer/build/resolve.go:74]: you could set `d.unresolvedImages` to empty array instead of calling `RemoveFromSlice` here.

**Review by tejal29 — APPROVED** [review]
I would prefer to describe the `--force` in this PR so change as well as documentation for the change  go hand in hand. 

Leaving it upto you.

**Comment by MarlonGamez** [comment]
@tejal29 Added a section to the docs, PTAL. I wasn't sure exactly where to place it within the init page

**Comment by tejal29** [comment]
> @tejal29 Added a section to the docs, PTAL. I wasn't sure exactly where to place it within the init page

Looks good! Ship it!

File: pkg/skaffold/initializer/build/resolve.go (hunk 1 above)
Comment (gsquared94) [reply pkg/skaffold/initializer/build/resolve.go]: sorry I missed this comment. A map would have been more efficient for this `RemoveFromSlice` behavior.

**Pull request merged**